src/path.cr

```diff
@@ -0,0 +1,847 @@
+  #
+  # If the path is empty or `"."`, it returns `"."`. If the path is rooted
+  # and in the top-most hierarchy, the root path is returned.
+  #
+  # ```
+  # Path["foo/bar/file.cr"].parent # => Path["foo/bar"]
+  # Path["foo"].parent             # => Path["."]
+  # Path["/foo"].parent            # => Path["/"]
+  # Path["/"].parent               # => Path["/"]
+  # Path[""].parent                # => Path["."]
+  # Path["foo/bar/."].parent       # => Path["foo/bar"]
+  # ```
+  def parent : Path
+    new_instance dirname
+  end
+
+  # Returns all parent paths of this path beginning with the topmost path.
+  #
+  # ```
+  # Path["foo/bar/file.cr"].parents # => [Path["."], Path["foo"], Path["foo/bar"]]
+  # ```
+  def parents : Array(Path)
+    parents = [] of Path
+    each_parent do |parent|
+      parents << parent
+    end
+    parents
+  end
+
+  # Yields each parent of this path beginning with the topmost parent.
+  #
+  # ```
+  # Path["foo/bar/file.cr"].each_parent { |parent| puts parent }
+  # # Path["."]
+  # # Path["foo"]
+  # # Path["foo/bar"]
+  # ```
+  def each_parent(&block : Path ->)
+    return if @name.empty? || @name == "."
+
+    first = true
+    @type.each_part_separator_index(@name) do |pos|
+      if pos == 0 || (pos == 2 && @name[1] == ':')
+        first = false
+        break if pos == @name.bytesize - 1 || @name.byte_slice(pos + 1).each_char.all? { |char| @type.separators.includes?(char) || char == '.' }
+        path = anchor || new_instance(@type.separator_string)
+      else
+        if first && @name[0] != '.'
+          yield new_instance "."
+        end
+        first = false
+
+        break if pos == @name.bytesize - 1
+        path = new_instance @name.byte_slice(0, pos)
+      end
+
+      yield path
+    end
+
+    if first
+      # this path didn't contain any separators
+      yield new_instance "."
+    end
+  end
+
+  # Returns the last component of this path.
+  #
+  # ```
+  # Path["/foo/bar/file.cr"].basename # => "file.cr"
+  # Path["/foo/bar/"].basename        # => "bar"
+  # Path["/"].basename                # => "/"
+  # Path[""].basename                 # => ""
+  # ```
+  def basename : String
+    return "" if @name.bytesize == 0
+    return @name if @type.separator_string?(@name)
+
+    reader = Char::Reader.new(at_end: @name)
+    while @type.separators.includes?(reader.current_char)
+      reader.pos -= 1
+    end
+
+    return @name[0..0] if reader.pos == -@name.bytesize - 1
+
+    if index = @type.last_index_of_separator(@name, reader.pos)
+      @name.byte_slice(index + 1, reader.pos - index)
+    else
+      @name.byte_slice(0, reader.pos + 1)
+    end
+  end
+
+  # Returns the last component of this path with *suffix* stripped from the
+  # end.
+  #
+  # ```
+  # Path["/foo/bar/file.cr"].basename(".cr") # => "file"
+  # Path["/foo/bar/file"].basename(".cr")    # => "file"
+  # ```
+  def basename(suffix : String) : String
+    basename.chomp(suffix)
```

Comment:
**Nitpicking**: simple implementation but allocates 2 strings and throws one away. It's acceptable and can be optimized later, but `#extension` does just the same, so the common `File.extname(path, File.extname(path))` ends up allocating 4 strings when we expected 2 (and executed `#basename` twice).